### PR TITLE
Explicit fallthrough

### LINKED ISSUE #2440: Layout fallthrough
### Describe the bug

I am trying to throw an error:

```js
export async function load({ page }) {
  if (page.params.foo !== 'okay') {
    return
  }
  return {}
}
```

This example works as expected within a page component. However when I put this logic inside `__layout.svelte` it behaves rather oddly.

In first case when nothing is returned then application proceeds to the page component. It renders the page contents, however ignores the __layout content. When the page is refreshed and reloaded completelly it thows an error 500 saying "load must return a value except for page fall through"

I want to check path validity for all the child pages within one directory and putting this logic inside every single page is not ideal. Better solution would to be able to check only once, inside the layout component.

I understand this is probably outside of the scope of intended functionality but it might be a mice addition, in my opinion.

### Reproduction

https://github.com/lemmon/svelte-kit-layout-error

### Logs

_No response_

### System Info

```shell
System:
    OS: macOS 11.5.2
    CPU: (4) x64 Intel(R) Core(TM) i5-6500 CPU @ 3.20GHz
    Memory: 49.27 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.17.1 - /usr/local/bin/node
    npm: 7.22.0 - /usr/local/bin/npm
  Browsers:
    Brave Browser: 88.1.19.86
    Chrome: 93.0.4577.82
    Firefox: 92.0
    Safari: 14.1.2
  npmPackages:
    @sveltejs/kit: next => 1.0.0-next.168 
    svelte: ^3.34.0 => 3.42.6
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by nhe23**
This PR makes fallthrough explicit. Only if `{fallthrough: true}` is returned by a load function or an endpoint the fallthrough mechanism is triggered. This way it is more clear when fallthrough is applied.
Fixes #2440 by allowing the same fallthrough logic that is used for page components. 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by Rich-Harris** [comment]
Thank you. My first reaction was that this could be confusing — it's much more likely that you'd simply forget to return something from `load` than that you'd want a layout to fall through.

But you could easily argue that the same is true for pages and endpoints. I've definitely been caught out by that once or twice (i.e. 404s that I wasn't expecting). 

So I was wondering if we shouldn't make it explicit — make the return value required (i.e. `load` and endpoint handlers _must_ return something), and make fallthroughs explicit:

```svelte
<script context="module">
  async function load({ params }) {
    if (params.foo !== 'okay') {
      return { fallthrough: true };
    }
    return {};
  }
</script>
```

Thoughts?

**Comment by nhe23** [comment]
The implicit fallthrough when nothing is returned is indeed a bit confusing. I'll change that to your proposed solution. When `fallthrough` is returned all other properties won't be considered. Should this be reflected in the return type of load and endpoints? So for load either the normal load output is returned or `{fallthrough:true}`. The type could look like this:

```ts 

export interface LoadOutput<
	Props extends Record<string, any> = Record<string, any>,
	Stuff extends Record<string, any> = Record<string, any>
> {
	status?: number;
	error?: string | Error;
	redirect?: string;
	props?: Props;
	stuff?: Stuff;
	maxage?: number;
} | {fallthrough: boolean }

```

Or is it enough to point this behavior out in the docs?

**Comment by dummdidumm** [comment]
It definitely should be reflected in the types 👍

**Comment by Rich-Harris** [comment]
It should be `| { fallthrough: true }` rather than `| { fallthrough: boolean }` I think (there's no reason you'd have `false`), but other than that yeah. Similarly for `RequestHandler`: https://github.com/sveltejs/kit/blob/a59c89fcbda510c386dc06d5a92a381c38b5132f/packages/kit/types/endpoint.d.ts#L12-L18

**Comment by nhe23** [comment]
Fallthrough is now explicit as suggested by @Rich-Harris. In order to reflect this in the types I added an `Either` type to the helpers, because with [Union Types](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#defining-a-union-type) only the values that the types have in common are directly accessible. So it would have been necessary to first check the returned type before accessing returned properties.

**Review by Rich-Harris — COMMENTED** [review]
Thanks, this is looking good — just a couple of small things I found

File: documentation/docs/01-routing.md [hunk 1]
```diff
@@ -79,15 +79,19 @@ export interface EndpointOutput<Body extends DefaultBody = DefaultBody> {
...
 	Locals = Record<string, any>,
 	Input = unknown,
 	Output extends DefaultBody = DefaultBody
 > {
 	(request: Request<Locals, Input>):
-		| void
+		| Fallthrough
 		| EndpointOutput<Output>
-		| Promise<void | EndpointOutput<Output>>;
+		| Promise<Fallthrough | EndpointOutput<Output>>;
```
[the remark this answers is not in the document]
Comment (nhe23) [reply documentation/docs/01-routing.md]: Added MaybePromise in latest commit

**Review by Rich-Harris — APPROVED** [review]

**Review by Rich-Harris — APPROVED** [review]

**Comment by faysal247** [comment]
To marge it in bot requests

**Pull request merged**